File: AdventOfCode2022/Day02/Day02.hs
import Data.List.Split (splitWhen)
import Data.List (sort)

elfToInt x
  | x == 'A' = 1 
  | x == 'B' = 2 
  | x == 'C' = 3 
  | otherwise = error "Unexpected"

meToInt x
  | x == 'X' = 1 
  | x == 'Y' = 2 
  | x == 'Z' = 3 
  | otherwise = error "Unexpected"

score1 [elf, _, me]
  | a == b = b + 3
  | b - a == 1 = b + 6
  | b == 1 && a == 3 = b + 6
  | otherwise = b
  where
    a = elfToInt elf
    b = meToInt me
score1 x = error "invalid input"

meToInt2 elf me
  | me == 'X' = if elf == 1 then 3 else elf - 1
  | me == 'Y' = elf
  | otherwise = if elf == 3 then 1 else elf + 1

score2 [elf, _, me]
  | a == b = b + 3
  | b - a == 1 = b + 6
  | b == 1 && a == 3 = b + 6
  | otherwise = b
  where
    a = elfToInt elf
    b = meToInt2 a me
score2 x = error "invalid input"

main = do
  content <- readFile "input.txt"
  -- part 1
  let totalScore1 = sum . map score1 $ lines content  
  print totalScore1

  -- part 2
  let totalScore2 = sum . map score2 $ lines content
  print totalScore2

File: AdventOfCode2022/Day02/day02.js
const fs = require("fs");

const data = fs.readFileSync("Day02/input.txt", "utf8");

const rounds = data.split("\n").map((x) => x.trim().split(" "));

const elf = {
  A: 1,
  B: 2,
  C: 3,
};

const me = {
  X: 1,
  Y: 2,
  Z: 3,
};

const score1 = (x, y) => {
  const a = elf[x];
  const b = me[y];
  if (a == b) {
    return b + 3;
  } else if ((a == 3 && b == 1) || b - a == 1) {
    return b + 6;
  }

  return b;
};

const scores1 = rounds.map(([a, b]) => score1(a, b));

const total1 = scores1.reduce((r, v) => r + v, 0);
console.log(total1);

const meToInt = (elf, me) => {
  if (me == "X") return elf == 1 ? 3 : elf - 1;
  if (me == "Y") return elf;
  return elf == 3 ? 1 : elf + 1;
}

const score2 = (x, y) => {
  const a = elf[x];
  let b = meToInt(a, y);
  if (a == b) {
    return b + 3;
  } else if ((a == 3 && b == 1) || b - a == 1) {
    return b + 6;
  }

  return b;
};
const scores2 = rounds.map(([a, b]) => score2(a, b));

const total2 = scores2.reduce((r, v) => r + v, 0);
console.log(total2);
